.github/workflows/codeql.yml

```diff
@@ -0,0 +1,59 @@
+name: CodeQL
+
+on:
+  push:
+    branches: [ master ]
+  pull_request:
+    branches: [ master ]
+
+jobs:
+  codeql:
+    runs-on: ubuntu-22.04
+    steps:
+    - name: Maximize build space
+      uses: easimon/maximize-build-space@master
+      with:
+        root-reserve-mb: 20480 # reserve for CodeQL
+        remove-dotnet: 'true'
+        remove-android: 'true'
+        remove-haskell: 'true'
+        remove-docker-images: 'true'
+
+    - name: Install ubuntu environment
+      shell: bash
+      run: |
+        export DEBIAN_FRONTEND=noninteractive
+        sudo apt-get update
+        sudo apt-get install -y git wget rpm rpm2cpio cpio make build-essential binutils m4 libtool-bin libncurses5
+
+    - uses: actions/checkout@v3
+
+    - name: Cache deps
+      id: cache-deps
+      uses: actions/cache@v3
+      env:
+        cache-name: cache-deps
+      with:
+        key: ${{ runner.os }}-build-${{ env.cache-name }}-el9.x86_64-${{ hashFiles('deps/init/oceanbase.el9.x86_64.deps') }}
+        path: deps/3rd
+
+    - name: Initialize CodeQL
+      uses: github/codeql-action/init@v2
+      with:
+        languages: cpp
+
+    - name: Build init
+      run: |
+        bash build.sh init
+
+    - name: Build project
+      shell: bash
+      run: |
+        rm -rf build_debug
+        bash build.sh debug
+        cd build_debug && make -j4 && cd -
```

Comment:
Does the build action can use ccache?

Reply:
> Does the build action can use ccache?

no, CodeQL requires a complete build to take place in order to perform analysis. ref: https://docs.github.com/en/code-security/code-scanning/troubleshooting-code-scanning/no-source-code-seen-during-build

Reply:
Sorry, can you explain the words 'a complete build`?
I didn't find some information about it.
And another question, what's the difference between build without cache and build with cache to CodeQL?

Reply:
Here is an interesting information:
https://codeql.github.com/docs/codeql-overview/supported-languages-and-frameworks/.
I think that CodeQL would use compiler to extract some information about the source code, so it requires `a complete build`.